### PR TITLE
blesh: init module

### PR DESCRIPTION
**Pull request by laalsaas**
###### Description of changes
create a module to properly source blesh in bashrc instead of using the `$(blesh-share)` workaround.
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aiotter** [comment]
Nice work! How about adding config?
FYR we have a module on home-manager which is still a draft: nix-community/home-manager#3238

**Comment by SuperSandro2000** [comment]
@ofborg eval

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 1]
```diff
@@ -74,6 +74,7 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - [appvm](https://github.com/jollheef/appvm), Nix based app VMs. Available as [virtualisation.appvm](options.html#opt-virtualisation.appvm.enable).
 
+- [blesh](https://github.com/akinomyoga/ble.sh), a line editor written in pure bash. Avilable as [programs.bash.blesh](#opt-programs.bash.blesh.enable).
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2211.section.md:77]:
```suggestion
- [blesh](https://github.com/akinomyoga/ble.sh), a line editor written in pure bash. Avilable as [programs.bash.blesh](#opt-programs.bash.blesh.enable).

```

File: nixos/modules/programs/bash/blesh.nix [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  cfg = config.programs.bash.blesh;
+in {
+  options = {
+    programs.bash.blesh.enable = mkEnableOption (mdDoc "blesh");
+  };
+
+  config = {
+    programs.bash.interactiveShellInit = optionalString cfg.enable mkMerge [ (mkBefore ''
+      source ${pkgs.blesh}/share/ble.sh --noattach
+    '') (mkAfter ''
```
Comment (SuperSandro2000) [inline nixos/modules/programs/bash/blesh.nix:13]:
```suggestion
    '')
    (mkAfter ''
```

File: nixos/doc/manual/release-notes/rl-2211.section.md (hunk 1 above)
Comment (laalsaas) [reply nixos/doc/manual/release-notes/rl-2211.section.md:77]: What change do you suggest? i see no difference. I can only guess you meant to fix Avilable -> Available

Comment (laalsaas) [reply nixos/doc/manual/release-notes/rl-2211.section.md:77]: And you introduced a blank line. didn't spot that at first sight.

**Comment by laalsaas** [comment]
> > How about adding config?
> 
> The config is full bash code, so it must be extraConfig.

However, iirc, there is no such thing as a global bleshrc. So should blesh config just go into bashrc? that doesn't seem clean to me either.

**Comment by laalsaas** [comment]
Also, I don't know why ofborg is failing.

**Comment by aiotter** [comment]
You can take the same approach I took at the home-manager module.
I mean adding `--rcfile=/path/to/rcfile` argument to `source ble.sh`.

https://github.com/nix-community/home-manager/blob/9e8d03d4d4d728c8128328f62c04209d7a6e1ae3/modules/programs/blesh.nix#L111-L118


@SuperSandro2000 Are you suggesting not to add any config options other than `programs.bash.blesh.extraConfig`?

File: nixos/doc/manual/release-notes/rl-2211.section.md (hunk 1 above)
Comment (SuperSandro2000) [reply nixos/doc/manual/release-notes/rl-2211.section.md]: GitHub can't display that properly...

**Comment by SuperSandro2000** [comment]
> Also, I don't understand why ofborg is failing.

you need to rebase on master

> Are you suggesting not to add any config options other than `programs.bash.blesh.extraConfig`?

What else should we do? Copy all the options from the wiki? Then we should rather do a generic style setting where we couldn't allow more complex shell things.

> Also, with using blesh like this, my PS1 customizations in `~/.bashrc` aren't respected.

That works for me when setting PS* between sourcing and attaching.

**Comment by laalsaas** [comment]
Since I'm the most competent git user in this universe, I fucked up the rebase, didn't realise it until I force pushed a branch that has 0 unique commits. I opened a new PR #202369, where I also already updated the release notes section, since branch-off happened yesterday.

**Comment by aiotter** [comment]
@SuperSandro2000 
>> Are you suggesting not to add any config options other than programs.bash.blesh.extraConfig?

> What else should we do? Copy all the options from the wiki? Then we should rather do a generic style setting where we couldn't allow more complex shell things.

Thank you for adding description. Please do not take my question more than a question. My mother tongue is not English, so sometimes I have difficulty in taking full meanings of your words.

**Comment by ShamrockLee** [comment]
@laalsaas I did it twice before.

`git rebase upstream/master` isn't always trustworthy, and the subsequent  `git push` would end the world. :sob:

`git rebase --onto upstream/master feature~n feature` would be safer.

**Pull request closed without merging**